django/applications/catmaid/static/js/widgets/tagtable.js

```diff
@@ -0,0 +1,369 @@
+/* -*- mode: espresso; espresso-indent-level: 8; indent-tabs-mode: t -*- */
+/* vim: set softtabstop=2 shiftwidth=2 tabstop=2 expandtab: */
+
+(function(CATMAID) {
+
+  "use strict";
+
+  var CACHE_TIMEOUT = 5*60*1000;  // cache invalidation timeout in ms
+
+  var TagTable = function() {
+    this.widgetID = this.registerInstance();
+    this.selectedSkeletons = new CATMAID.BasicSkeletonSource(this.getName());
+  };
+
+  $.extend(TagTable.prototype, new InstanceRegistry());
+
+  TagTable.prototype.getName = function() {
+    return 'Tag Table ' + this.widgetID;
+  };
+
+  var labelSkelMappingCache = {};  // todo - button for enabling/disabling cache?
+
+  /**
+   *
+   * @param labelIDs - array of label IDs
+   * @param callback - function which takes a list of skelIDs
+   */
+  var processSkelIDsFromLabelIDs = function(labelIDs, callback) {
+    var remoteLabelIDs = [];
+    var skelIDs = new Set();
+
+    var now = Date.now();
+
+    for (var i = 0; i < labelIDs.length; i++) {
+      var labelID = labelIDs[i];
+      if (labelID in labelSkelMappingCache && now - labelSkelMappingCache[labelID].timestamp <= CACHE_TIMEOUT) {
+        skelIDs.addAll(
+          labelSkelMappingCache[labelID].skelIDs
+        );
+      } else {
+        remoteLabelIDs.push(labelID);
+      }
+    }
+
+    if (remoteLabelIDs.length === 0) {
+      callback(Array.from(skelIDs));
+      return;
+    }
+
+    CATMAID.fetch(project.id + '/skeletons/node_labels', 'POST', {
+      'label_ids': remoteLabelIDs
+    }).then(function(json) {
+      now = Date.now();
+
+      for (var i = 0; i < json.length; i++) {
+        var labelSkelTuple = json[i];
+        labelSkelMappingCache[labelSkelTuple[0]].skelIDs = labelSkelTuple[1];
+        labelSkelMappingCache[labelSkelTuple[0]].timestamp = now;
+        skelIDs.addAll(labelSkelTuple[1]);
+      }
+
+      callback(Array.from(skelIDs));  // todo - replace with promise?
```

Comment:
I'm probably not going to have time to do this before I leave - I had a read of the docs earlier it looked like a bit much to get my head around among the other changes at that point - a constructor which takes a function which takes two arguments which are both functions and so on. Happy to go that way in future, though, and see what it looks like if one of you guys could finish it off for this PR.